src/metabase/driver/postgres/actions.clj

```diff
@@ -3,52 +3,58 @@
   (:require
    [clojure.java.jdbc :as jdbc]
    [metabase.driver.sql-jdbc.actions :as sql-jdbc.actions]
+   [metabase.driver.sql-jdbc.connection :as sql-jdbc.conn]
+   [metabase.util :as u]
    [metabase.util.i18n :refer [tru]]))
 
-(defn- constraint->columns [conn constraint-name]
-  (->> ["select column_name from information_schema.constraint_column_usage where constraint_name = ?" constraint-name]
-       (jdbc/query conn {:identifers identity, :transaction? false})
-       (map :column_name)))
-
-(defn- violates-not-null-constraint [_conn error-message]
+(defn- maybe-parse-not-null-error [_database error-message]
   (let [[match? value column]
         (re-find #"ERROR:\s+(\w+) value in column \"([^\"]+)\" violates not-null constraint" error-message)]
     (when match?
       [{:message (tru "{0} violates not-null constraint" value)
         :column column}])))
 
-(defn- violates-unique-constraint [conn error-message]
+(defn- constraint->column-names
+  "Given a constraint with `constraint-name` fetch the column names associated with that constraint."
+  [database constraint-name]
+  (let [jdbc-spec (sql-jdbc.conn/db->pooled-connection-spec (u/the-id database))
+        sql-args  ["select column_name from information_schema.constraint_column_usage where constraint_name = ?" constraint-name]]
+    (into []
+          (map :column_name)
+          (jdbc/reducible-query jdbc-spec sql-args {:identifers identity, :transaction? false}))))
```

Comment:
`reducible-query` isn't needed here but it's slightly more memory-efficient and works nicely with `(into [] ...)` which I wanted to use here